test/system/src/test/java/io/pravega/test/system/ByteClientTest.java

```diff
@@ -0,0 +1,176 @@
+        ensureSegmentStoreRunning(zkUri, controllerUri);
+    }
+
+    @Before
+    public void setup() {
+        Service conService = Utils.createPravegaControllerService(null);
+        List<URI> ctlURIs = conService.getServiceDetails();
+        controllerURI = ctlURIs.get(0);
+        streamManager = StreamManager.create(controllerURI);
+        assertTrue("Creating scope", streamManager.createScope(SCOPE));
+        assertTrue("Creating stream", streamManager.createStream(SCOPE, STREAM, config));
+    }
+
+    @After
+    public void tearDown() {
+        streamManager.close();
+        ExecutorServiceHelpers.shutdown(executor);
+        ExecutorServiceHelpers.shutdown(writerExecutor);
+        ExecutorServiceHelpers.shutdown(readerExecutor);
+    }
+
+    /**
+     * This test verifies the correctness of basic read/write functionality of {@link ByteStreamClient}.
+     */
+    @Test
+    public void byteClientTest() throws IOException {
+        @Cleanup
+        ConnectionFactory connectionFactory = new ConnectionFactoryImpl(ClientConfig.builder().build());
+        ControllerImpl controller = new ControllerImpl(ControllerImplConfig.builder()
+                .clientConfig(ClientConfig.builder()
+                        .controllerURI(controllerURI).build()).build(),
+                connectionFactory.getInternalExecutor());
+        @Cleanup
+        ClientFactory clientFactory = new ClientFactoryImpl(SCOPE, controller);
+        log.info("Invoking byteClientTest test with Controller URI: {}", controllerURI);
+
+        ByteStreamClient byteStreamClient = clientFactory.createByteStreamClient();
+        @Cleanup("closeAndSeal")
+        ByteStreamWriter writer = byteStreamClient.createByteStreamWriter(STREAM);
+        @Cleanup
+        ByteStreamReader reader = byteStreamClient.createByteStreamReader(STREAM);
+
+        for (int i = 1; i <= MAX_PAYLOAD_SIZE; i *= 10) {
+            final int payloadSize = i;
+            // Create the synthetic payload for the write.
+            byte[] payload = new byte[payloadSize];
+            byte[] readBuffer = new byte[payloadSize];
+            RandomFactory.create().nextBytes(payload);
+            final int payloadHashCode = Arrays.hashCode(payload);
+            log.info("Created synthetic payload of size {} with hashcode {}.", payload.length, payloadHashCode);
+            AtomicInteger writerIterations = new AtomicInteger();
+            AtomicInteger readerIterations = new AtomicInteger();
+
+            // Write the same synthetic payload multiple times to the Stream.
+            CompletableFuture<Void> writerLoop = Futures.loop(() -> writerIterations.get() < IO_ITERATIONS,
+                    () -> CompletableFuture.runAsync(() -> {
+                        try {
+                            log.debug("Writing payload of size: {}. Iteration {}.", payload.length, writerIterations.get());
+                            writer.write(payload);
+                            writer.flush();
```

Comment:
It might be worth covering the case where there is more than one write inflight

Reply:
Now the test considers writes with and without flushing.